### DISCUSSION TITLE
How to know how many layers of adaptivity have been applied at a specific quadrature point?

### DISCUSSION THREAD

**Discussion by Minjiang-Zhu** [idaholab/moose#26131] [Q&A General] [answered]
Hi, I want to apply different levels of adaptivity w.r.t. different variables. Say, if I want to apply 5 layers of adaptivity to variable u and 3 layers of adaptivity to variable v by ValueRange marker, I will need to stop marking v as 'refine' if 3 layers of adaptivity are already applied. I will modify the source file 'ValueRangeMarker.C' by myself, yet how should I know the adaptivity level?

**Comment by GiudGiud** [comment]
Hello

Do you mean the current local adaptivity level for an element?
Can you use this to combine two markers?
https://mooseframework.inl.gov/source/markers/ComboMarker.html

Guillaume

**Reply by Minjiang-Zhu** [reply]
Hi, I know how to combine them, yet the maximize level of refinement is defined by 'max_h_level'. I want to apply different maximum levels to these two markers, yet I don't know how to obtain the current level.

**Reply by GiudGiud** [reply, marked as answer]
for a given element, you can get the level of adaptivity with elem->level()
see the doxygen
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1Elem.html#a9c0282a545257463ba99b1cbb7f9752d

then in a custom marker you can have logic that checks the level before marking

**Reply by lindsayad** [reply]
It sounds like you want a different refinement level for each variable? That is not possible. The level of refinement is stored on a per-element basis and applies uniformly to the variables in the equation systems. There is a capability to do selective p-refinement, but even there you can only totally disable p-refinement for a variable; there is no option to only (p-)refine a variable up to a certain amount

**Reply by Minjiang-Zhu** [reply]
Hi @lindsayad, I actually want to apply different max_h_level to two markers. I think @GiudGiud just provided me what I want. Thank you both! I will try to implement it now.

**Reply by lindsayad** [reply]
excellent!
